File: revdecode/update_ranking_scheme.py
import json

'''
This script is used to update the ranking scheme. The original ranking scheme is like:
[[1], [2,2,2], [3,3], [4], [5]], the number here represent the rank number.
The new ranking scheme is like:
[[1], [2,2,2], [5,5], [7], [8]].
The main purpose for this change is to do detailed performance analysis based on NDCG metric.
'''


def extract_ranking(data_list):
    """
    Extracts the ranking format from the data list.

    Args:
    - data_list: List[Dict[str, Union[str, int]]].
      A list of dictionaries containing 'rank' and potentially other attributes.

    Returns:
    - List[List[int]]. The intermediate format of ranks.
    """
    rank_list = [item['rank'] for item in data_list]
    intermediate_ranking = []
    while rank_list:
        value = rank_list.pop(0)
        sublist = [value]
        while rank_list and rank_list[0] == value:
            sublist.append(rank_list.pop(0))
        intermediate_ranking.append(sublist)
    return intermediate_ranking


def transform_ranking(intermediate_ranking):
    """
    Transforms the intermediate ranking to the new ranking format.

    Args:
    - intermediate_ranking: List[List[int]]. The intermediate format of ranks.

    Returns:
    - List[List[int]]. The transformed ranks in intermediate format.
    """
    new_ranking = []
    prev_rank = 0
    for sublist in intermediate_ranking:
        new_rank = prev_rank + 1
        new_ranking.append([new_rank] * len(sublist))
        prev_rank = new_rank + len(sublist) - 1
    return new_ranking


def create_new_data_with_rank(data_list, transformed_ranking):
    """
    Creates a new list with the transformed ranking and the original data.

    Args:
    - data_list: List[Dict[str, Union[str, int]]]. The original list with unmodified ranks.
    - transformed_ranking: List[List[int]]. The ranks in transformed format.

    Returns:
    - List[Dict[str, Union[str, int]]]. New data list with transformed ranks.
    """
    new_data = []
    data_index = 0
    for sublist in transformed_ranking:
        for rank in sublist:
            updated_item = data_list[data_index].copy()
            updated_item['rank'] = rank
            new_data.append(updated_item)
            data_index += 1
    return new_data


def updating_ranking_for_one_target_function(ranking_info):
    """
    Updates the ranking for one target function.

    Args:
    - ranking_info: List[Dict[str, Union[str, int]]]. The original ranking information.
    """
    intermediate = extract_ranking(ranking_info)
    transformed = transform_ranking(intermediate)
    updated_ranking = create_new_data_with_rank(ranking_info, transformed)

    return updated_ranking

File: revdecode/src/update_ranking_scheme.cpp
// update_ranking_scheme.cpp
#include "update_ranking_scheme.h"

using namespace std;
using json = nlohmann::json;
using ordered_json = nlohmann::ordered_json;
using VariantType = std::variant<std::string, int, double>;

// Function to extract the ranking information from the data list
vector<vector<int>> extract_ranking(const vector<map<string, VariantType>> &data_list) {
    // Create a flat list of all ranks
    vector<int> rank_list;
    for (const auto &item : data_list) {
        rank_list.push_back(get<int>(item.at("rank")));
    }

    // Grouping consecutive equal ranks into sublists
    vector<vector<int>> intermediate_ranking;
    while (!rank_list.empty()) {
        int value = rank_list.front();
        rank_list.erase(rank_list.begin());
        vector<int> sublist = {value};
        while (!rank_list.empty() && rank_list.front() == value) {
            sublist.push_back(rank_list.front());
            rank_list.erase(rank_list.begin());
        }
        intermediate_ranking.push_back(sublist);
    }

    return intermediate_ranking;
}

// Function to handle the ties in ranking
// Example input: [[1], [2, 2, 2], [3, 3], [4], [5]]
// Example output: [[1], [2, 2, 2], [5, 5], [7], [8]]
vector<vector<int>> transform_ranking(const vector<vector<int>> &intermediate_ranking) {
    vector<vector<int>> new_ranking;
    int prev_rank = 0;
    for (const auto &sublist : intermediate_ranking) {
        int new_rank = prev_rank + 1;
        new_ranking.push_back(vector<int>(sublist.size(), new_rank));
        prev_rank = new_rank + static_cast<int>(sublist.size()) - 1;
    }

    return new_ranking;
}

// Function to update the data list with the new_ranking
vector<map<string, VariantType>> create_new_data_with_rank(const vector<map<string, VariantType>> &data_list, const vector<vector<int>> &transformed_ranking) {
    vector<map<string, VariantType>> new_data;
    size_t data_index = 0;
    for (const auto &sublist : transformed_ranking) {
        for (int rank : sublist) {
            map<string, VariantType> updated_item = data_list[data_index];
            updated_item["rank"] = rank;
            new_data.push_back(updated_item);
            data_index++;
        }
    }
    return new_data;
}

// Given the original ranking information, handle the ties by using the helper functions defined
vector<map<string, VariantType>> update_ranking_for_one_target_function(const vector<map<string, VariantType>> &ranking_info) {
    auto intermediate = extract_ranking(ranking_info);
    auto transformed = transform_ranking(intermediate);
    auto updated_ranking = create_new_data_with_rank(ranking_info, transformed);
    return updated_ranking;
}

// Function to convert json to vector of maps
vector<map<string, VariantType>> json_to_vector_map(const ordered_json &j) {
    vector<map<string, VariantType>> vec;
    for (const auto &item : j) {
        map<string, VariantType> map;
        for (const auto &element : item.items()) {
            if (element.value().is_string()) {
                map[element.key()] = element.value().get<string>();
            } else if (element.value().is_number_integer()) {
                map[element.key()] = element.value().get<int>();
            } else if (element.value().is_number_float()) {
                double original_value = element.value().get<double>();
                map[element.key()] = original_value;
            }
        }
        vec.push_back(map);
    }
    return vec;
}

// Function to convert vector of maps to json
ordered_json vector_map_to_json(const vector<map<string, VariantType>> &vec) {
    ordered_json j = ordered_json::array();

    for (const auto &map : vec) {
        ordered_json map_json;

        // Add attributes in the specific order
        if (map.find("library_name") != map.end()) {
            if (holds_alternative<string>(map.at("library_name"))) {
                map_json["library_name"] = get<string>(map.at("library_name"));
            }
        }
        if (map.find("library_version") != map.end()) {
            if (holds_alternative<string>(map.at("library_version"))) {
                map_json["library_version"] = get<string>(map.at("library_version"));
            }
        }
        if (map.find("compiler_option") != map.end()) {
            if (holds_alternative<string>(map.at("compiler_option"))) {
                map_json["compiler_option"] = get<string>(map.at("compiler_option"));
            }
        }
        if (map.find("function_name") != map.end()) {
            if (holds_alternative<string>(map.at("function_name"))) {
                map_json["function_name"] = get<string>(map.at("function_name"));
            }
        }
        // Add compiler_unit
        if (map.find("compiler_unit") != map.end()) {
            if (holds_alternative<string>(map.at("compiler_unit"))) {
                map_json["compiler_unit"] = get<string>(map.at("compiler_unit"));
            }
        }
        if (map.find("similarity_score") != map.end()) {
            if (holds_alternative<double>(map.at("similarity_score"))) {
                double value = get<double>(map.at("similarity_score"));
                map_json["similarity_score"] = value;
            }
        }
        // Add emission_probability
        if (map.find("confidence_factor") != map.end()) {
            if (holds_alternative<double>(map.at("confidence_factor"))) {
                double value = get<double>(map.at("confidence_factor"));
                map_json["confidence_factor"] = value;
            }
        }
        // Add uniqueness_score
        if (map.find("uniqueness_score") != map.end()) {
            if (holds_alternative<double>(map.at("uniqueness_score"))) {
                double value = get<double>(map.at("uniqueness_score"));
                map_json["uniqueness_score"] = value;
            }
        }
        if (map.find("rank") != map.end()) {
            if (holds_alternative<int>(map.at("rank"))) {
                map_json["rank"] = get<int>(map.at("rank"));
            }
        }
        if (map.find("original_rank") != map.end()) {
            if (holds_alternative<int>(map.at("original_rank"))) {
                map_json["original_rank"] = get<int>(map.at("original_rank"));
            }
        }

        j.push_back(map_json);
    }

    return j;
}